### PR TITLE
Improve management of many workspace thumbnails

### LINKED ISSUE #5: 21.10: picker previews scale strangely when many workspaces are created
This is a lot more apparent with the new picker styling:
Screenshot from 2021-10-07 10-03-43

but it does still happen without the styling applied:
Screenshot from 2021-10-07 10-38-46

The top preview also starts to touch the top edge of the picker box as more workspaces get created, and that happens before the picker box is filled up.

### PR DESCRIPTION
**Pull request by 13r0ck**
Closes #5 by scaling Left/Right Offset to snap scaling to fit the nearest number of workspaces that fit the screen. Then adds scrolling to workspaces to show an effectively unlimited number of workspaces without shrinking workspaces thumbnails to the point they become meaningless.
> Improvement on #22 by scaling doesn't rely on magic numbers, and add scroll.

https://user-images.githubusercontent.com/58987761/144342354-f6fab32e-102c-4dd8-a881-676ca58f06f5.mp4

# ~~Todo:~~
- ~~Drag-and-drop windows should scroll workspaces, so users can move windows to workspaces that are hidden by scroll.~~
- ~~Indicator to show that there are more workspaces hidden by scroll. @maria-komarova How should I implement this?~~ final functionality is different, and doesn't need these

### CODE REVIEW

File: workspaceThumbnail.js
Comment (jackpot51) [inline workspaceThumbnail.js]: Could you remove this log message?

Comment (13r0ck) [reply workspaceThumbnail.js]: Ah for sure. Thanks!

**Comment by jacobgkau** [comment]
This seems to be working pretty well, although I think the scroll indicator (mentioned in the `Todo` section of the PR description) would be necessary to ship this. It can get a little confusing navigating through the list without it, since I can't exactly scroll the list without changing the active desktop.

I thought I saw a different workspace from the one I clicked get selected once or twice earlier, but I haven't been able to recreate that yet.

**Comment by maria-komarova** [comment]
Trying to think through this, my question is what the behavior would be for dynamic workspaces when one is dragging windows to a new workspace, and then the new workspace is created automatically. Would it show outside of the frame with the indicator appearing, or would it "push" other thumbnails up?

**Comment by 13r0ck** [comment]
My thinking for was that holding a window below the gray box should scroll down through workspaces. And holding it above the box would do the opposite. Then once the user sees the workspace they want they can hover inside the gray box, which then stops scrolling, and then they could hover and drop onto the workspace they want.

**Comment by maria-komarova** [comment]
Where would the dynamically created workspace show?

**Comment by 13r0ck** [comment]
In the current release, there is not indicator if the user drops their window above or below the gray box. It just auto-appends/prepends. 

The dynamic workspace is always displayed at the end of the list. As far as I can tell there is no way to disable that via extension. Hide-able yes, but removable no.

**Comment by maria-komarova** [comment]
What I mean is that where would it show if thumbnails could be scrolled when it is created - inside the box or outside?

**Comment by maria-komarova** [comment]
If inside, then it results in the workspace thumbnails above it getting automatically scrolled up, so to speak. If outside, then how does one know a workspace was created automatically? If with the indicator, then would it be enough to tell what has just happened?

**Comment by maria-komarova** [comment]
And initiating dragging, then hovering over something while waiting for it to scroll, and only then dropping the window onto a workspace could be a challenging experience. Either the scrolling could be too slow or too fast. 
So while having larger thumbnails could be beneficial, I'm not sure the overall experience would be better than scaling like in 21.04.

**Comment by 13r0ck** [comment]
OOooh, gotcha. 
Umm, I didn't think about that. Maybe if we add the "this list is scrollable" indicator it could have an animation "eating" the workspace if it is created outside of the visible list?

**Comment by 13r0ck** [comment]
Though on a regular 1080p monitor a user will have 8~9 workspaces visible before scrolling even begins. I doubt most users will need more than that.

But with 10~15 workspaces they get so small that having to perfectly align the mouse within a 10px rectangle isn't a great UX either

**Comment by maria-komarova** [comment]
Right, it's not great. But it might still be preferable. On 21.04 the thumbnails start scaling down after 8th workspace is added. With 10 thumbnails it still feels okay to drag and drop. So if we have similar size thumbnails, we'll only fit 7 in the 1080 display.
Drag and drop when the target area needs to be scrolled to (or panned to) could result in a lot of accidental mistakes, primarily because of the speed of scroll and needing to move cursor away from the hover area that triggers the scroll at the right moment to get to the needed thumbnail. Or if it is too slow - then a lot of annoyance. It can also result in more disorientation.

**Review by jackpot51 — DISMISSED** [review]

**Comment by 13r0ck** [comment]
After conversation here and privately with maria, it was decided that scrolling shouldn't be a part of this PR due to potentially bad UX when dragging and dropping windows into other workspaces. As such, scrolling was removed and now this PR just improves workspace thumbnail scaling from ~5 to ~20 workspaces. Scaling still breaks at >20 workspaces on a 1080p monitor, though at that point, the thumbnails have gotten too small to be useful. This should still be good enough to close #5 until maria is available to design a better system for workspace management.

https://user-images.githubusercontent.com/58987761/144928531-43e31806-9cc2-4e1f-a8fb-d3e57f89f922.mp4

**Review by jackpot51 — DISMISSED** [review]

**Review by leviport — APPROVED** [review]
@13r0ck has some ideas on how to make this even better, but they'll probably take time to implement. Until then, I think it's worth merging this since it does improve the behavior quite a bit. Future improvements can be made in a separate PR.

**Review by leviport — CHANGES_REQUESTED** [review]
Actually don't merge this one. It causes a regression, presumably related to https://github.com/pop-os/cosmic-workspaces/issues/16 which makes the previews not scale at all and spill off the bottom of the screen when a 4k display connected.

**Comment by leviport** [comment]
Screenshot of the issue I described above:
Screenshot from 2021-12-08 16-51-04
There are about 4 workspaces off the bottom edge of both screens

**Comment by 13r0ck** [comment]
Trying the same PR on an external 1080p monitor I do not get the same bug. Cosmic-workspaces appears to handle hidpi environments poorly

**Comment by 13r0ck** [comment]
:partying_face: I think I finally did it! Tested with multi monitors and fixes the bug levi had above. I also increased number of thumbnails that can be shown on a 1080p monitor to 100+

https://user-images.githubusercontent.com/58987761/145662963-115168f7-40ee-4f26-ab3c-820b8b2f538a.mp4

**Review by jackpot51 — APPROVED** [review]

**Review by leviport — APPROVED** [review]
This looks great! Tested up to around 50 workspaces, and it handled them very well.

**Pull request merged**
